arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -464,19 +464,23 @@ static bool sync_shadow_pasid_table_entry(struct pgt_sync_data *sdata)
 	if (type == PASID_ENTRY_PGTT_FL_ONLY) {
 		struct pkvm_pgtable_cap cap;
 
-		/*
-		 * When host IOMMU driver is using first-level only
-		 * translation, pkvm IOMMU will actually use nested
-		 * translation to add one more layer translation to
-		 * guarantee the protection. This one more layer is the
-		 * EPT.
-		 */
-		type = PASID_ENTRY_PGTT_NESTED;
+		if (ptdev_attached_to_vm(ptdev))
```

Comment:
This assumes that the host has already issued PKVM_HC_ADD_PTDEV hypercall before, i.e. that the VMM should do KVM_DEV_VFIO_GROUP_ADD before mapping the guest memory in vIOMMU via VFIO_IOMMU_MAP_DMA, not after it, right?
Crosvm doesn't follow this assumption, so I had to implement https://chromium-review.googlesource.com/c/crosvm/crosvm/+/4194734 to make it work correctly with crosvm.

Reply:
If PKVM_HC_ADD_PTDEV hypercall is already called, then if (ptdev_attached_to_vm(ptdev)) is true, otherwise it will be false. So both cases can be handled. And this function is to set the pasid entry, so it doesn't rely on if the virtual second-level page table (maintained by host VM) has been mapped or not (via VFIO_IOMMU_MAP_DMA) because the physical second-level page table (pgstate_pgt) is synced with the virtual second-level page table right before handling the first EPT_VIOLATION, not here. Did I misunderstand your point?

Reply:
Right, the mapping itself is done not here but at the first EPT violation, but here we rely on ptdev_attached_to_vm(ptdev) for configuring which translation type and which page tables to use. So if PKVM_HC_ADD_PTDEV has not been already called, ptdev_attached_to_vm(ptdev) is still false and ptdev->pgt is still the host EPT. So if we don't get here again after PKVM_HC_ADD_PTDEV is finally called, the translation will be still set up so that the device will have access to the host VM memory and not to the protected VM memory (whereas it should be the other way around), so DMA in the protected VM will not work.

That's what I'm actually observing with crosvm: when it starts the VM, it first setups the IOMMU, then calls PKVM_HC_ADD_PTDEV and then never setups the IOMMU again. To be precise, I'm actually testing it with legacy mode (since I have no HW with scalable mode), with my https://github.com/intel-staging/pKVM-IA/pull/17/commits/3f201ad27e15b6ece8738070f61dbe9bf29c2aed, but it's similar to scalable mode in this regard. In my case with https://github.com/intel-staging/pKVM-IA/pull/17/commits/3f201ad27e15b6ece8738070f61dbe9bf29c2aed this issue actually doesn't cause non-working DMA, it just causes using directly the host's virtual tables instead of pgstate_pgt (which is also rather wrong, but the device DMA in the protected VM actually happens to work), but it's only thanks to the temporary workaround for legacy mode. In scalable mode (and in legacy mode after we finally implement shadow page tables), as I understand, it will not work at all in this case, as I mentioned above.

Reply:
When host VMM uses PKVM_HC_ADD_PTDEV, pKVM handles this with calling pkvm_iommu_sync() after setting ptdev->shadow_vm_handle and ptdev->pgt. pkvm_iommu_sync() should call sync_shadow_pasid_table_entry() and  ptdev_attached_to_vm(ptdev) is true. Do you mean that when pkvm is handling PKVM_HC_ADD_PTDEV, the pkm_iommu_sync() doesn't call sync_shadow_pasid_table_entry() and finally cause the IOMMU not setup?

Reply:
Hmm indeed, pkvm_iommu_sync() should ensure proper IOMMU setup, forgot about that. I need to double-check, maybe there was a bug in my code.

Reply:
I've checked, with my newest patches pkvm_iommu_sync() indeed ensures correct IOMMU setup. Sorry for confusion.
IOW, what I wrote in NOTE: in https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/4161812/3 is not quite correct.